src/main/java/edu/cornell/kfs/fp/service/impl/AmazonWebServicesBillingServiceImpl.java

```diff
@@ -0,0 +1,636 @@
+        return costCenterDTO;
+    }
+    
+    private void updateTransactionDTOChartAccountSubAccount(AmazonBillingDistributionOfIncomeTransactionDTO transactionDTO, AmazonBillingCostCenterDTO costCenterDTO,
+            String amazonDetailKFSAccount) {
+        Account account;
+        if (StringUtils.isNotBlank(costCenterDTO.getAccountNumber())) {
+            account = buildAccount(costCenterDTO.getAccountNumber());
+        } else {
+            account = buildAccount(amazonDetailKFSAccount);
+        }
+        if (validAccount(account)) {
+            transactionDTO.setAccountNumber(account.getAccountNumber());
+            updateTansactionDTOChart(transactionDTO, costCenterDTO, account);
+            updateTransactionDTOSubAccount(transactionDTO, costCenterDTO, account);
+        } else {
+            LOG.info("updateTransactionDTOChartAccountSubAccount() Invalid account.");
+            transactionDTO.setTransactionInputError(true);
+        }
+    }
+    
+    private Account buildAccount(String accountNumber) {
+        if (StringUtils.isNotBlank(accountNumber)) {
+            Account account = getAccountService().getUniqueAccountForAccountNumber(accountNumber);
+            return account;
+        }
+        return null;
+    }
+    
+    private boolean validAccount(Account account) {
+        return account != null && account.isActive() && !account.isExpired();
+    }
+    
+    private void updateTansactionDTOChart(AmazonBillingDistributionOfIncomeTransactionDTO transactionDTO, AmazonBillingCostCenterDTO costCenterDTO, 
+            Account account) {
+        if (StringUtils.isNotBlank(costCenterDTO.getChartCode())) {
+            Chart chart = getChartService().getByPrimaryId(costCenterDTO.getChartCode());
+            if (ObjectUtils.isNotNull(chart) && chart.isActive()) {
+                transactionDTO.setChartCode(chart.getChartOfAccountsCode());
+            } else {
+                LOG.info("updateTansactionDTOChart() Invalid Chart.");
+                transactionDTO.setTransactionInputError(true);
+            }
+        } else {
+            transactionDTO.setChartCode(account.getChartOfAccountsCode());
+        }
+    }
+    
+    private void updateTransactionDTOSubAccount(AmazonBillingDistributionOfIncomeTransactionDTO transactionDTO, AmazonBillingCostCenterDTO costCenterDTO, 
+            Account account) {
+        if (StringUtils.isNotBlank(costCenterDTO.getSubAccountNumber())) {
+            SubAccount subAccount = buildSubAccount(account, costCenterDTO.getSubAccountNumber());
+            if (ObjectUtils.isNotNull(subAccount)) {
+                transactionDTO.setSubAccountNumber(subAccount.getSubAccountNumber());
+            } else {
+                LOG.info("updateTransactionDTOSubAccount() Invalid Sub Account.");
+                transactionDTO.setTransactionInputError(true);
+            }
+        }
+    }
+    
+    private SubAccount buildSubAccount(Account account, String subAccountNumber) {
```

Comment:
this method again looks like it is only retrieving the sub account rather then building it from input data, probably would be good to rename the method